packages/@lwc/template-compiler/src/codegen/codegen.ts

```diff
@@ -456,6 +459,12 @@ export default class CodeGen {
         }
     }
 
+    genDynamicEventListeners(onDirective: OnDirective) {
+        const onDirectiveArg = this.bindExpression(onDirective.value);
+        const freezedArg = this._renderApiCall(RENDER_APIS.freeze, [onDirectiveArg]);
+        return t.property(t.identifier('dynamicOn'), freezedArg);
```

Comment:
Indirect AST building is hard to parse. Can you add a comment with the expected result?